From cf2f686bf5df64f1ded27c70d541f450d95faad5 Mon Sep 17 00:00:00 2001
From: Yvan Volochine <yvan.volochine@gmail.com>
Date: Sat, 20 Jul 2013 16:08:21 +0200
Subject: [PATCH] scide: multi-editor: add shortcuts to be able to jump to a
 tab

- Alt+1 through Alt+8 jumps to the document at specified position
- Alt+9 jumps to last document
---
 editors/sc-ide/widgets/multi_editor.cpp | 47 +++++++++++++++++++++++++++++++++
 editors/sc-ide/widgets/multi_editor.hpp |  1 +
 2 files changed, 48 insertions(+)

diff --git a/editors/sc-ide/widgets/multi_editor.cpp b/editors/sc-ide/widgets/multi_editor.cpp
index 2eee368..67c1601 100644
--- a/editors/sc-ide/widgets/multi_editor.cpp
+++ b/editors/sc-ide/widgets/multi_editor.cpp
@@ -234,6 +234,48 @@ class DocumentSelectPopUp : public QDialog
     setCurrentBox( defaultBox ); // will updateActions();
 
     applySettings( main->settings() );
+
+
+    // A signal mapper to avoid creating custom slots for each tab
+    QSignalMapper *mapper = new QSignalMapper(this);
+
+    // Shortcuts for tabs (1 to 8)
+    QShortcut *s1 = new QShortcut(QKeySequence("Alt+1"), this);
+    connect( s1, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s1, 0);
+
+    QShortcut *s2 = new QShortcut(QKeySequence("Alt+2"), this);
+    connect( s2, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s2, 1);
+
+    QShortcut *s3 = new QShortcut(QKeySequence("Alt+3"), this);
+    connect( s3, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s3, 2);
+
+    QShortcut *s4 = new QShortcut(QKeySequence("Alt+4"), this);
+    connect( s4, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s4, 3);
+
+    QShortcut *s5 = new QShortcut(QKeySequence("Alt+5"), this);
+    connect( s5, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s5, 4);
+
+    QShortcut *s6 = new QShortcut(QKeySequence("Alt+6"), this);
+    connect( s6, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s6, 5);
+
+    QShortcut *s7 = new QShortcut(QKeySequence("Alt+7"), this);
+    connect( s7, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s7, 6);
+
+    QShortcut *s8 = new QShortcut(QKeySequence("Alt+8"), this);
+    connect( s8, SIGNAL(activated()), mapper, SLOT(map()) );
+    mapper->setMapping(s8, 7);
+    // Last tab
+    QShortcut *s9 = new QShortcut(QKeySequence("Alt+9"), this);
+    connect( s9, SIGNAL(activated()), this, SLOT(showLastDocument()) );
+
+    connect( mapper, SIGNAL(mapped(int)), mTabs, SLOT(setCurrentIndex(int)) );
 }
 
 void MultiEditor::makeSignalConnections()
@@ -862,6 +904,11 @@ void MultiEditor::showPreviousDocument()
     mTabs->setCurrentIndex( qMax(0, currentIndex - 1) );
 }
 
+void MultiEditor::showLastDocument()
+{
+    mTabs->setCurrentIndex( mTabs->count() - 1 );
+}
+
 void MultiEditor::switchDocument()
 {
     CodeEditorBox *box = currentBox();
diff --git a/editors/sc-ide/widgets/multi_editor.hpp b/editors/sc-ide/widgets/multi_editor.hpp
index 943b2be..21101c9 100644
--- a/editors/sc-ide/widgets/multi_editor.hpp
+++ b/editors/sc-ide/widgets/multi_editor.hpp
@@ -132,6 +132,7 @@ class MultiEditor : public QWidget
 
     void showNextDocument();
     void showPreviousDocument();
+    void showLastDocument();
     void switchDocument();
 
     void splitHorizontally() { split(Qt::Horizontal); }
-- 
1.8.4
